### PR TITLE
Types - get 'runtime.connect()' to return a 'runtime.Port'

### PR DESCRIPTION
**Pull request by mckenfra**
See issue: https://github.com/stoically/webextensions-api-mock/issues/1
See stubs pr: https://github.com/stoically/webextensions-api-mock/pull/3 (and old types pr: https://github.com/stoically/webextensions-api-mock/pull/2)

This pull request is for the updates to the types in the `types.d.ts` file, to go along with the pull request mentioned above for the stubs.

As discussed on the other pull requests, I'm not sure how to test this. The generation works, and the build works without errors, but I don't have an IDE to test that the typescript interfaces/types are all now lined up correctly with the stubs!

### CODE REVIEW

**Comment by mckenfra** [comment]
Just to let you know, I'm currently rewriting this to use a common `SchemaWalker` class, shared with the `stubs` code. So maybe hold off your review for the time being.

Also I'm not totally happy with the `types.d.ts` file produced by this PR. I'm going to try to get it to produce similar output to the existing version on `master`, but with `runtime.connect()` returning a `runtime.Port`. Also the `enum` properties don't look right to me currently - e.g. `windows.WindowType` is an array: `WindowsWindowType[]`, but shouldn't the type be `WindowsWindowType` for this property?

**Comment by stoically** [comment]
> Just to let you know, I'm currently rewriting this to use a common SchemaWalker class, shared with the stubs code. So maybe hold off your review for the time being.

Great, thanks! I'll keep an eye on the PR then.

> I'm going to try to get it to produce similar output to the existing version on master, but with runtime.connect() returning a runtime.Port

Sounds good.

> Also the enum properties don't look right to me currently - e.g. windows.WindowType is an array: WindowsWindowType[], but shouldn't the type be WindowsWindowType for this property?

You're right, they're not correct. I didn't notice because the main consumer of this package, [webextensions-api-fake](https://github.com/stoically/webextensions-api-fake/), doesn't fake constant properties (yet), so yeah, good catch!

**Comment by mckenfra** [comment]
Ok I've finished the rewrite - the `stubs` and `types` code now share a common `SchemaWalker` class. Unfortunately this solution isn't quite as simple as I'd hoped - there's quite a lot of changed code, so may be difficult to review! Maybe I can try to simplify and reduce the code at some point. What might be good would be for you to check that the new generated `types.d.ts` output is correct: https://github.com/stoically/webextensions-api-mock/blob/5b8d566526fb44988b1a0f21b5b78981b1c54858/src/generated/types.d.ts

File: src/generated/types.d.ts [hunk 1]
```diff
@@ -1020,82 +1267,101 @@ export type PrivacyWebsitesTrackingProtectionModeOption =
...
+  Port: {
+    disconnect: sinon.SinonStub;
+    name: string;
+    onDisconnect: Events['Event'];
+    onMessage: Events['Event'];
+    postMessage: sinon.SinonStub;
+    sender?: Runtime['MessageSender'];
+  };
+  RequestUpdateCheckStatus: RuntimeRequestUpdateCheckStatus;
+  connect: () => Runtime['Port'];
```
Comment (stoically) [inline src/generated/types.d.ts]: I'd expect this to be an sinon stub interface that also returns the Runtime['Port'], so something like

```ts
interface SinonStubReturns<T> extends sinon.SinonStub {
  () => T;
}

interface Runtime {
  connect: SinonStubReturns<Runtime['Port']>
}
```

That allows asserting calls to `connect()` and changing `connect`s behavior by e.g. calling `connect.returns()` - and I believe it's what the stub generator already does?

**Review by stoically — COMMENTED** [review]
This looks much more sophisticated and correct, I can only imagine how tricky that was!

File: src/generated/types.d.ts (hunk 1 above)
Comment (mckenfra) [reply src/generated/types.d.ts]: Ok sure, sounds good. I'm pretty unfamiliar with `Sinon`, so I definitely welcome your help / guidance regarding what `Sinon` types to use. Yes the `stubs` generator calls `stub().returns(returnValue)`. But I wasn't sure how to convert this to a typescript `interface` property. I also couldn't find where `sinon.SinonStub` is defined. I've looked in `node_modules/sinon` but can't find any definitions for `SinonStub`. Any ideas?

Regarding your `SinonStubReturns` interface - does that also already exist? Or would we need to add it to the `types.d.ts` file?

Comment (stoically) [reply src/generated/types.d.ts]: > Yes the stubs generator calls stub().returns(returnValue)

Right, thanks for confirming.

> couldn't find where sinon.SinonStub is defined

The [sinon types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/sinon/ts3.1/index.d.ts) come from [`@types/sinon`](https://github.com/stoically/webextensions-api-mock/blob/fa55f0e235c480ecf6ab1f0eeddceb00848956f6/package.json#L49). 

> Regarding your SinonStubReturns interface - does that also already exist? Or would we need to add it to the types.d.ts file?

That doesn't exist, and adding it like that should be sufficient, yeah.

Comment (mckenfra) [reply src/generated/types.d.ts]: Aha, I've found it: `node_modules/\@types/sinon/index.d.ts`

Comment (mckenfra) [reply src/generated/types.d.ts]: Ok I've now changed this in https://github.com/stoically/webextensions-api-mock/pull/4/commits/b0058ba823df856f0470e41c9e8dac287c5aa195 as follows:

```typescript
interface Runtime {
    connect: sinon.SinonStub<any[], Runtime['Port']>;
}
```

This is based on the types I found in `node_modules/\@types/sinon/ts3.1/index.d.ts`. And it seems to build ok!

**Comment by stoically** [comment]
Looks great now. Would you mind rebasing on master and adding `@types/chai`, `@types/sinon-chai` as devdependencies?

**Comment by stoically** [comment]
Oh, and adding `generate` at the end of the `build` script would be appreciated too, so it runs in CI.

**Comment by mckenfra** [comment]
> Oh, and adding `generate` at the end of the `build` script would be appreciated too, so it runs in CI.

You mean something like this in `package.json`?

```javascript
"scripts": {
  "build": "npm-run-all lint test build:*",
  "build:tsc": "tsc -p tsconfig.build.json",
  "build:generate": "ts-node src/bin.ts",
  "build:copy": "copyup \"src/generated/*\" dist",
  "generate": "ts-node src/bin.ts",
  /* etc. */
}
```

File: package.json [hunk 2]
```diff
@@ -14,6 +14,7 @@
     "build": "npm-run-all lint test build:*",
```
Comment (stoically) [inline package.json]:
```suggestion
     "build": "npm-run-all lint test build:* generate",
```

Comment (mckenfra) [reply package.json]: Cool, sure

**Comment by stoically** [comment]
Scratch that. I'll add it to the CI manually, since it might not be desired to generate when building.

**Comment by mckenfra** [comment]
> Scratch that. I'll add it to the CI manually, since it might not be desired to generate when building.

ok cool - I was going to say if you add it to the end of the `build` command like you suggested, then the`build:copy` step would have been copying possibly out-of-date files from `src/generated` to `dist/generated`...

**Comment by stoically** [comment]
Right.

(Btw, I didn't meant to delete your comment, it was timing, apparently I hit delete the moment you commented; I wanted to delete my suggestion, should've just commented again, duh)

**Comment by mckenfra** [comment]
> Right.
> 
> (Btw, I didn't meant to delete your comment, it was timing, apparently I hit delete the moment you commented; I wanted to delete my suggestion)

Haha - ok cool. Now rebased and added those dev dependencies: https://github.com/stoically/webextensions-api-mock/pull/4/commits/a0b07f55513872ea6d9692ab3dc211c4c53b0ea4

**Review by stoically — APPROVED** [review]
I'm really happy with this change, thanks a ton!

**Comment by stoically** [comment]
Will have to test this a bit from consuming packages before cutting a release, I'll try to do it soon!

**Comment by stoically** [comment]
Everything looking good! Released as v1.0.0 :tada:

**Comment by mckenfra** [comment]
Fantastic, thanks a lot! :smiley: :thumbsup:

**Pull request merged**